pandas/_libs/index.pyx

```diff
@@ -270,7 +270,14 @@ cdef class IndexEngine:
             Py_ssize_t i, j, n, n_t, n_alloc
 
         self._ensure_mapping_populated()
-        values = np.array(self._get_index_values(), copy=False)
+        if any([checknull(t) for t in targets]):
```

Comment:
This already exists as a function called `isnaobj` in missing.pyx - can you use that instead?

Reply:
I changed 273 to `if isnaobj(targets).any():` and tested against the `index_cached_properties`, `index_object`, `indexing`, and `indexing_engines` benchmarks and got the following results:

```
     <master>         <nui-regression>
+         172±1μs          247±3μs     1.44  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int16Engine'>, <class 'numpy.int16'>), 'monotonic_incr')
+         188±2μs         245±30μs     1.30  index_cached_properties.IndexCache.time_is_monotonic('PeriodIndex')
+         210±2μs          237±2μs     1.13  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int8Engine'>, <class 'numpy.int8'>), 'monotonic_incr')
+      13.3±0.3μs       14.7±0.6μs     1.11  index_cached_properties.IndexCache.time_shape('CategoricalIndex')
-        11.3±2μs       10.0±0.3μs     0.89  index_cached_properties.IndexCache.time_is_all_dates('Float64Index')
-        249±30μs         215±10μs     0.86  index_cached_properties.IndexCache.time_is_monotonic('TimedeltaIndex')
-        18.8±2μs       15.9±0.9μs     0.85  index_cached_properties.IndexCache.time_engine('UInt64Index')
-      1.10±0.2ms        894±100μs     0.81  index_cached_properties.IndexCache.time_engine('MultiIndex')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Any suggestions to improve these?

Reply:
you need to create a mask then you don't actually check things multiple times.

why are you creating targets/values as object arrays?

Reply:
Makes sense. `targets[isnaobj(targets)].size > 0` would work, correct? Because now instead of iterating through the array with `any()`, I'm just filtering the given array and checking its size.

I created the targets/values arrays as object arrays because I thought that it was possible to have multiple dtypes in the given arrays. Thinking about it now, I figure it makes more sense to make them as their previous array dtypes:

```
targets = np.array(new_targets, dtype=targets.dtype)
values = np.array(new_values, dtype=self._get_index_values().dtype)
```

I implemented this and the tests are passing. Does this sound good? Any suggestion to improve?

Reply:
@WillAyd I ran tests against indexing and indexing_engines benchmarks. No performance regressions:

```
asv continuous -f 1.1 upstream/master HEAD -b ^indexing_engines
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

asv continuous -f 1.1 upstream/master HEAD -b indexing
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```